### PR TITLE
Clear only determination when COType is changing

### LINKED ISSUE #5198: Adding a Collecting Event record to a CO deletes all Determinations
**Describe the bug**
Adding a Collecting Event record on the CO deletes all Determinations

**To Reproduce**

1. Go to a record with a 'Collection Object Type' set (`/specify/view/collectionobject/126/?recordsetid=13`)
2. Add a dependent Collecting Event via the subview
3. See the determination(s) are all deleted

**Expected behavior**
Those should not be deleted!

**Screenshots**

https://github.com/user-attachments/assets/b28bfa20-e4e0-4fe8-9a6e-26d68d8f9feb

**System Information:**
Arc, macOS

Running locally, this is similar to the database `geo_swiss_2024_08_05` on the test panel.

GitSHA: https://github.com/specify/specify7/commit/196fe71f3ba3276b6fb766b1ed3a338362b01408

[Specify 7 System Information - 2024-08-07T14_46_34.262Z.txt](https://github.com/user-attachments/files/16530336/Specify.7.System.Information.-.2024-08-07T14_46_34.262Z.txt)

### LINKED ISSUE #5199: Form Clearing when COType is changing
Currently when the user is changing the COType on a CO form, all the fields are cleared. 

New behavior requested: 
No fields are being cleared expect for the Determinations

### PR DESCRIPTION
**Pull request by CarolineDenis**
Fixes #5199
Fixes #5198

### Checklist

- [x] Self-review the PR after opening it to make sure the changes look good
      and self-explanatory (or properly documented)
- [x] Add automated tests
- [x] Add relevant issue to release milestone

### Testing instructions

Fixes #5199
- Open CO from 
- Add COType 
- Add Determination 
- Add other fields 
- Save 
- Change the COType 
- [ ] Verify Determination has been erased but other fields have kept their value 

Fixes #5198
- Open CO from 
- Add COType 
- Add Determination 
- Save
- Add CE
- [ ] Verify Determination has not been erased

### CODE REVIEW

**Review by emenslin — APPROVED** [review]
### Testing instructions

- Open CO from 
- Add COType 
- Add Determination 
- Add other fields 
- Save 
- Change the COType 
- [x] Verify Determination as been erased but other fields have kept their value 

Looks good, only determinations are removed when COType is changed

**Review by lexiclevenger — APPROVED** [review]
### Testing instructions

- Open CO from 
- Add COType 
- Add Determination 
- Add other fields 
- Save 
- Change the COType 
- [x] Verify Determination as been erased but other fields have kept their value 
---
Looks good! Only fields from the Determination table are cleared.

https://github.com/user-attachments/assets/2c7b0a4b-8b98-42f1-9324-4b53dd7747b5

**Review by alesan99 — CHANGES_REQUESTED** [review]
### Testing instructions

Fixes #5199
- Open CO from 
- Add COType 
- Add Determination 
- Add other fields 
- Save 
- Change the COType 
- [x] Verify Determination as been erased but other fields have kept their value 


Fixes #5198

- Go to a record with a 'Collection Object Type'
- Add a determination if not present 
- Save 
- Add a dependent Collecting Event via the subview
- [ ] See the determination(s) are not deleted

Issue #5199 is fixed but #5198 is still happening to me.
Seems to only happen after a refresh, not when immediately adding a CE after saving.

https://github.com/user-attachments/assets/3e3808a7-1ed3-47a7-bd9a-b8d22c6556c8

EDIT: #5198 is no longer in the scope of this PR, so its working as expected 👍

**Review by alesan99 — APPROVED** [review]

File: specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts [hunk 1]
```diff
@@ -160,40 +159,41 @@ export const businessRuleDefs: MappedBusinessRuleDefs = {
...
+        const taxon = determinations?.models[0].get('taxon');
+        if (typeof taxon !== 'string') return;
+        const taxonId = idFromUrl(taxon ?? '');
+        const COType = resource.get('collectionObjectType');
+        const COTypeId = idFromUrl(COType);
+
+        const fetchTaxon = fetchResource('Taxon', taxonId ?? 0);
+        const fetchCOType = fetchResource(
+          'CollectionObjectType',
+          COTypeId ?? 0
```
Comment (CarolineDenis) [inline specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts:176]: For taxonId ?? 0  and COTypeId ?? 0 , we can just handle the case when they are null or undefined by returning.

**Review by sharadsw — APPROVED** [review]

**Review by melton-jason — CHANGES_REQUESTED** [review]
Functionally, this is fine as long as this is acceptable for the requirements. I'd love to see a test for the clearing functionality! 😃

Building a little more on the requirements, I will caution that if it is _never_ acceptable to have determinations on a CollectionObject which are associated to taxon records from a different taxon tree, then this is not a complete solution. 
Whether intentional or not, a user can still have determinations which have taxons that do not exist on the collectionObjectType -> taxontreedef.

File: specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts [hunk 2]
```diff
@@ -216,46 +217,50 @@ export const businessRuleDefs: MappedBusinessRuleDefs = {
...
+                .all({
+                  defaultType:
+                    collectionObject.get('collectionObjectType') === undefined
+                      ? collectionObject
+                          .rgetPromise('collection')
+                          .then(async (collection) =>
+                            collection.rgetPromise('collectionObjectType')
+                          )
+                          .then((coType) => coType ?? undefined)
+                      : undefined,
```
Comment (sharadsw) [inline specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts:231]: Maybe we need a `customInit` as well for CoType which sets the default cotype in a CO form? Could extract this logic into a function and use it there. What do you think?

Comment (CarolineDenis) [reply specifyweb/frontend/js_src/lib/components/DataModel/businessRuleDefs.ts:231]: we can create a new issue to set default cotype :+1:

**Review by combs-a — APPROVED** [review]
### Testing instructions
Fixes #5199
- Open CO from 
- Add COType 
- Add Determination 
- Add other fields 
- Save 
- Change the COType 
- [X] Verify Determination has been erased but other fields have kept their value 

https://github.com/user-attachments/assets/b31d60db-b629-4a20-b0e3-05851c59614a

Fixes #5198
- Open CO from 
- Add COType 
- Add Determination 
- Save
- Add CE
- [X] Verify Determination has not been erased

https://github.com/user-attachments/assets/72d92da8-6e98-436a-a07c-3f77d42fc18a

Looks good, is working!

**Pull request merged**